Lib/asyncio/tasks.py

```diff
@@ -522,60 +522,78 @@ def _on_completion(f):
         fut.remove_done_callback(cb)
 
 
-# This is *not* a @coroutine!  It is just an iterator (yielding Futures).
-def as_completed(fs, *, loop=None, timeout=None):
-    """Return an iterator whose values are coroutines.
+class as_completed(object):
+    """Asynchronous iterator over awaitables which returns the result of each
+    as completed.
 
-    When waiting for the yielded coroutines you'll get the results (or
-    exceptions!) of the original Futures (or coroutines), in the order
-    in which and as soon as they complete.
+    This differs from PEP 3148; results are yielded from asynchronous
+    iteration rather than futures.
```

Comment:
The new implementation should also return futures, just like the one from `concurrent.futures` does. The difference is that with async iteration we can return the _original_ futures, which are guaranteed to have completed.

Reply:
The existing implementation of `as_completed` already yields results. I also think yielding results rather than futures is more natural with `async for`.

Reply:
The existing implementation yields futures which you must `await`.

The idea is to yield futures which are guaranteed to have completed, as they complete. This is how `concurrent.futures.as_completed` works, but the benefit goes beyond consistency with `concurrent.futures`. It allows code to be robust in the face of exceptions raised by futures, and to associate side data with the futures and retrieve it from `as_completed`. For example, when yielding futures, [this usage example](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor-example) immediately becomes usable with `asyncio.as_completed`, just changing `for` to `async for`.

The sync version doesn't support that usage because it gives you a _different_ future than any of those passed to `as_completed`. Among other things, that's what prompted the BPO issue.

As far as I can tell, your implementation can easily support that, just by removing `.result()` at the `yield` point.

Reply:
You're right, I misread what was going on before. Thanks for the explanation!